### PR TITLE
Spec verbose debug reports for source registration

### PR DESCRIPTION
**Pull request by linnan-github**
***
<a href="https://pr-preview.s3.amazonaws.com/WICG/attribution-reporting-api/pull/613.html" title="Last updated on Nov 21, 2022, 6:18 PM UTC (d7cd970)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/attribution-reporting-api/613/14b657c...d7cd970.html" title="Last updated on Nov 21, 2022, 6:18 PM UTC (d7cd970)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -355,7 +357,20 @@ An attribution rate-limit record is a [=struct=] with the following items:
...
+
+A <dfn>trigger debug data type</dfn> is a [=debug data type=] for trigger registrations.
+Possible values include "`trigger-no-matching-source`", "`trigger-no-matching-filter-data`",
+"`trigger-attributions-per-source-destination-limit`", "`trigger-reporting-origin-limit`",
+"`trigger-event-deduplicated`", "`trigger-event-no-matching-configurations`",
+"`trigger-event-noise`", "`trigger-event-low-priority`", "`trigger-event-excessive-reports`",
+"`trigger-event-storage-limit`", "`trigger-event-outside-report-window`",
+"`trigger-aggregate-deduplicated`", "`trigger-aggregate-no-contributions`",
+"`trigger-aggregate-insufficient-budget`", "`trigger-aggregate-storage-limit`",
+"`trigger-aggregate-outside-report-window`", "`trigger-unknown-error`".
```
Comment (apasel422) [inline index.bs:373]: These would probably be easier to read if formatted as a `<ul>`.

File: index.bs [hunk 2]
```diff
@@ -939,6 +978,8 @@ To <dfn>process an attribution source</dfn> given an [=attribution source=] |sou
         1. [=set/Append=] |fakeReport| to the [=event-level report cache=].
     1. If |source|'s [=attribution source/randomized response=] is not [=set/is empty|empty=],
         then set |source|'s [=attribution source/event-level attributable=] value to false.
+    1. Run [=obtain and deliver a debug report on source registration=] with "`source-noised`" and |source|.
+    1. Return.
```
Comment (apasel422) [inline index.bs:982]: Just making sure this new `return` is intentional.

File: index.bs [hunk 3]
```diff
@@ -1236,15 +1282,87 @@ To <dfn>check if an [=attribution source=] can create [=aggregatable contributio
...
+    <dd>[=map/Set=] |body|["`limit`"] to [=max attribution reporting endpoints per rate-limit window=],
+         [=serialize an integer|serialized=].</dd>
+    <dt>"trigger-attributions-per-source-destination-limit"</dt>
+    <dd>[=map/Set=] |body|["`limit`"] to [=max attributions per rate-limit window=],
+         [=serialize an integer|serialized=].</dd>
+    <dt>"`trigger-event-low-priority`"</dt>
+    <dt>"`trigger-event-excessive-reports`"</dt>
+    <dd>
+        1. [=Assert=]: |report| is not null and is an [=event-level report=].
+        1. Return the result of running [=obtain an event-level report body=] with |report|.
```
Comment (apasel422) [inline index.bs:1318]: To be clear: We are returning the actual event-level report in this case?

File: index.bs [hunk 4]
```diff
@@ -1275,20 +1422,26 @@ To <dfn>trigger event-level attribution</dfn> given an [=attribution trigger=] |
     1. Let |matchingReports| be all entries in the [=event-level report cache=] where all of the following are true:
          * entry's [=event-level report/report time=] and |report|'s [=event-level report/report time=] are equal.
          * entry's [=event-level report/source identifier=] [=string/is=] |report|'s [=event-level report/source identifier=]
-    1. If |matchingReports| is empty, then set |sourceToAttribute|'s [=attribution source/event-level attributable=] value to false
-        and return <strong>dropped</strong>.
+    1. If |matchingReports| is empty:
+        1. Set |sourceToAttribute|'s [=attribution source/event-level attributable=] value to false.
+        1. Let |debugData| be the result of running [=obtain debug data on trigger registration=]
+            with "`trigger-event-excessive-reports`", |trigger|, |sourceToAttribute| and |report|.
```
Comment (apasel422) [inline index.bs:1428]: If there are no matching reports we want to return `excessive-reports`?

**Review by apasel422 — CHANGES_REQUESTED** [review]
Any chance we could split this into separate PRs, or at least separate commits within the PR, for each debug report type? It's quite hard to follow.

File: index.bs (hunk 2 above)
Comment (linnan-github) [reply index.bs:982]: Good catch, it's not.

File: index.bs (hunk 3 above)
Comment (linnan-github) [reply index.bs:1318]: Yes, the implementation is here (https://source.chromium.org/chromium/chromium/src/+/main:content/browser/attribution_reporting/attribution_debug_report.cc;l=296).

File: index.bs (hunk 4 above)
Comment (linnan-github) [reply index.bs:1428]: If there are no matching reports, then that means the number of event-level reports have been reached, and therefore we return `excessive-reports`.

**Comment by linnan-github** [comment]
> Any chance we could split this into separate PRs, or at least separate commits within the PR, for each debug report type? It's quite hard to follow.

Thanks Andrew. Yeah, let me try to break this down.

**Review by apasel422 — DISMISSED** [review]
Where do we check whether the debug cookie is set?

**Comment by linnan-github** [comment]
> Where do we check whether the debug cookie is set?

Debug cookie is not fully spec yet, but updated the algorithm to check that.

File: index.bs [hunk 5]
```diff
@@ -599,7 +599,7 @@ Given an [=attribution rate-limit record=] |record|:
 To <dfn>obtain and deliver a debug report</dfn> given a [=list=] of [=attribution debug data=] |data|
 and an [=origin=] |reportingEndpoint|:
 
-1. Assert that |reportingEndpoint| is a [=potentially trustworthy origin=].
+1. [=Assert =]: |reportingEndpoint| is a [=potentially trustworthy origin=].
```
Comment (apasel422) [inline index.bs:602]:
```suggestion
1. [=Assert=]: |reportingEndpoint| is a [=potentially trustworthy origin=].
```

File: index.bs [hunk 6]
```diff
@@ -895,10 +895,24 @@ a [=trigger state=] |triggerState|:
...
+To <dfn>check if debug reporting is allowed</dfn> given a [=source debug data type=] |dataType|
+and an [=origin=] |reportingOrigin|:
+1. If |dataType| is:
+    <dl class="switch">
+    <dt>"`source-destination-limit`"</dt>
+    <dd>Return <strong>allowed</strong>.</dd>
+    <dt>"`source-noised`"</dt>
+    <dt>"`source-storage-limit`"</dt>
+    <dt>"`source-unknown-error`"</dt>
+    <dd>Return t he result of running [=check if cookie-based debugging is allowed=] with |reportingOrigin|.</dd>
```
Comment (apasel422) [inline index.bs:907]:
```suggestion
    <dd>Return the result of running [=check if cookie-based debugging is allowed=] with |reportingOrigin|.</dd>
```

**Review by apasel422 — CHANGES_REQUESTED** [review]

**Review by apasel422 — APPROVED** [review]
